src/main/kotlin/com/terraformation/backend/accelerator/HubSpotService.kt

```diff
@@ -0,0 +1,538 @@
+import io.ktor.client.request.request
+import io.ktor.client.request.setBody
+import io.ktor.client.statement.HttpResponse
+import io.ktor.client.statement.bodyAsText
+import io.ktor.http.HttpMethod
+import io.ktor.http.HttpStatusCode
+import io.ktor.http.parameters
+import jakarta.inject.Named
+import jakarta.ws.rs.core.UriBuilder
+import java.math.BigDecimal
+import java.net.URI
+import java.net.URLEncoder
+import java.nio.charset.StandardCharsets
+import java.time.Instant
+import java.time.InstantSource
+import kotlinx.coroutines.runBlocking
+import org.jooq.DSLContext
+import org.springframework.security.access.AccessDeniedException
+
+@Named
+class HubSpotService(
+    private val clock: InstantSource,
+    private val config: TerrawareServerConfig,
+    private val countriesDao: CountriesDao,
+    private val dslContext: DSLContext,
+    private val httpClient: HttpClient,
+) {
+  companion object {
+    private const val ACCELERATOR_PIPELINE_LABEL = "Accelerator Projects"
+    private const val APPLICATION_STAGE_LABEL = "Application"
+
+    // HubSpot-defined properties
+    private const val COMPANY_PROPERTY_NAME = "name"
+    private const val COMPANY_PROPERTY_WEBSITE = "website"
+    private const val CONTACT_PROPERTY_EMAIL = "email"
+    private const val CONTACT_PROPERTY_FIRST_NAME = "firstname"
+    private const val DEAL_PROPERTY_NAME = "dealname"
+    private const val DEAL_PROPERTY_PIPELINE = "pipeline"
+    private const val DEAL_PROPERTY_STAGE = "dealstage"
+
+    // Custom properties
+    private const val CONTACT_PROPERTY_FULL_NAME = "full_name"
+    private const val DEAL_PROPERTY_APPLICATION_REFORESTABLE_LAND = "project_hectares"
+    private const val DEAL_PROPERTY_PROJECT_COUNTRY = "project_country"
+
+    // From https://developers.hubspot.com/docs/api/crm/associations#association-type-id-values
+    private const val ASSOCIATION_CATEGORY = "HUBSPOT_DEFINED"
+    private const val ASSOCIATION_TYPE_CONTACT_TO_DEAL = 4
+    private const val ASSOCIATION_TYPE_COMPANY_TO_DEAL = 342
+
+    private val hubSpotBaseUrl = URI("https://api.hubapi.com/")
+
+    private val log = perClassLogger()
+  }
+
+  private var accessToken: String? = null
+  private var accessTokenExpiration = Instant.EPOCH
+  private var acceleratorPipelineId: String? = null
+  private var applicationStageId: String? = null
+  private val dealPageUrlPrefix: URI by lazy { initDealPageUrlPrefix() }
+
+  private val redirectUri = config.webAppUrl.resolve("/admin/hubSpotCallback").toString()
+
+  private val projectCountryOptions: Set<String> by lazy {
+    sendRequest<GetProjectCountryPropertyResponse>(
+            "/crm/v3/properties/deals/$DEAL_PROPERTY_PROJECT_COUNTRY")
+        .options
+        .map { it.value }
+        .toSet()
+  }
+
+  /**
+   * Creates a deal and associates a contact and a company with it.
+   *
+   * @return The URL of the deal's page in the HubSpot UI.
+   */
+  fun createApplicationObjects(
```

Comment:
Would it be more straightforward if this function uses `ApplicationModel` and `ApplicationVariableValues` directly?